### DISCUSSION TITLE
Squaring the circle... or rather circling the square - a look at shaped colophons

### DISCUSSION THREAD

**Discussion by Omikhleia** [sile-typesetter/sile#1263] [Show and tell]
Just for the show :)

I have been working on a naive implementation of "paragraph shapes" (a.k.a. parshape's) in the (Knuth-Plass) line breaking algorithm. Not so naive, actually, with some tweaks to try an automated computation of the line widths depending on the content (Lua rocks here). Still fragile and easy to break, but I am getting _somewhere_:

![image](https://user-images.githubusercontent.com/18075640/138254918-ee8a4d2e-1392-4e08-bb05-756c4e442fdb.png)

**Comment by simoncozens** [comment]
Whoa. That's extremely awesome.

**Comment by alerque** [comment]
Yes, party in the house! I've manually cobbled together some shapes before but never cleaned it up to automate.

I assume you're setting up a module that returns the current line width (left/right margins) based on some geometry calculation of it's own and swapping that out in the line breaker?

**Comment by simoncozens** [comment]
TeX’s Knuth-Plass algorithm takes a “paragraph shape” parameter. I just didn’t implement that bit. Basically you give it an array of line lengths and instead of using a constant width, each time it finds a break it sets the target width for the next line to the next element of the array.

**Comment by Omikhleia** [comment]
Worth noting here, for casual readers, that the necessary features for such paragraph shaping fancyness are now part of SILE > v0.12 ;)
